File: backend/app/api/v1/store/transcripts.py
"""Transcript / segments / participants / chat read-only REST API.

These tables hang off a meeting. transcript_segments, meeting_participants and
meeting_chat_messages have NO org_id, so tenant isolation is enforced by first
resolving the parent meeting via ``scoped_meeting_or_404`` (which 404s if the
principal is not in the meeting's org) and only then querying by meeting_id.
"""

from __future__ import annotations

from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.v1.store._common import get_db, get_principal, scoped_meeting_or_404
from app.db.models import (
    MeetingChatMessage,
    MeetingParticipant,
    Transcript,
    TranscriptSegment,
)
from app.db.scope import Principal
from app.schemas.common import BaseSchema

router = APIRouter()


# ---- schemas --------------------------------------------------------------
class TranscriptResponse(BaseSchema):
    id: str
    full_text: str
    language: str
    word_count: int
    confidence_score: float | None = None
    created_at: datetime


class SegmentResponse(BaseSchema):
    id: str
    meeting_id: str
    speaker_label: str
    speaker_name: str | None = None
    text: str
    start_time: float
    end_time: float
    confidence: float | None = None
    segment_index: int
    is_partial: bool


class ParticipantResponse(BaseSchema):
    id: str
    meeting_id: str
    speaker_label: str
    speaker_name: str | None = None
    user_id: str | None = None
    email_address: str | None = None
    joined_at: datetime | None = None
    left_at: datetime | None = None


class ChatMessageResponse(BaseSchema):
    id: str
    meeting_id: str
    sender_name: str | None = None
    sender_email: str | None = None
    text: str
    sent_at: datetime


# ---- transcript -----------------------------------------------------------
@router.get("/meetings/{meeting_id}/transcript", response_model=TranscriptResponse)
def get_transcript(
    meeting_id: str,
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> TranscriptResponse:
    scoped_meeting_or_404(session, principal, meeting_id)
    transcript = session.scalar(
        select(Transcript).where(Transcript.meeting_id == meeting_id)
    )
    if transcript is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Transcript not found"
        )
    return TranscriptResponse.model_validate(transcript)


# ---- segments -------------------------------------------------------------
@router.get(
    "/meetings/{meeting_id}/transcript-segments",
    response_model=list[SegmentResponse],
)
def list_transcript_segments(
    meeting_id: str,
    speaker: str | None = Query(None),
    q: str | None = Query(None),
    limit: int = Query(200, ge=1, le=500),
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> list[SegmentResponse]:
    scoped_meeting_or_404(session, principal, meeting_id)
    stmt = select(TranscriptSegment).where(
        TranscriptSegment.meeting_id == meeting_id,
        TranscriptSegment.is_partial.is_(False),
    )
    if speaker:
        stmt = stmt.where(TranscriptSegment.speaker_label == speaker)
    if q:
        stmt = stmt.where(TranscriptSegment.text.ilike(f"%{q}%"))
    stmt = stmt.order_by(TranscriptSegment.start_time).limit(limit)
    rows = session.scalars(stmt).all()
    return [SegmentResponse.model_validate(r) for r in rows]


# ---- participants ---------------------------------------------------------
@router.get(
    "/meetings/{meeting_id}/participants",
    response_model=list[ParticipantResponse],
)
def list_participants(
    meeting_id: str,
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> list[ParticipantResponse]:
    scoped_meeting_or_404(session, principal, meeting_id)
    stmt = (
        select(MeetingParticipant)
        .where(MeetingParticipant.meeting_id == meeting_id)
        .order_by(MeetingParticipant.joined_at.is_(None), MeetingParticipant.joined_at)
    )
    rows = session.scalars(stmt).all()
    return [ParticipantResponse.model_validate(r) for r in rows]


# ---- chat -----------------------------------------------------------------
@router.get("/meetings/{meeting_id}/chat", response_model=list[ChatMessageResponse])
def list_chat(
    meeting_id: str,
    session: Session = Depends(get_db),
    principal: Principal = Depends(get_principal),
) -> list[ChatMessageResponse]:
    scoped_meeting_or_404(session, principal, meeting_id)
    stmt = (
        select(MeetingChatMessage)
        .where(MeetingChatMessage.meeting_id == meeting_id)
        .order_by(MeetingChatMessage.sent_at)
        .limit(500)
    )
    rows = session.scalars(stmt).all()
    return [ChatMessageResponse.model_validate(r) for r in rows]

File: worker/internal/store/transcripts.go
package store

import (
	"context"
	"database/sql"
	"errors"

	"github.com/Yrock1x/AI-Notetaker-Finance/worker/internal/model"
)

// scopedMeetingID resolves a meeting the principal's org owns, returning its id,
// else ErrNotFound (ports scoped_meeting_or_404: a missing row or one in a
// foreign org is a 404). The org_id filter is the tenant guard — transcript /
// segment / participant / chat rows hang off a meeting and (for segments and
// participants) carry no org_id of their own, so a missing scope here would
// silently leak cross-tenant data.
func scopedMeetingID(ctx context.Context, conn *sql.DB, p *Principal, meetingID string) error {
	pred, args := p.OrgFilter("org_id")
	q := "SELECT id FROM meetings WHERE id = ? AND " + pred
	var id string
	err := conn.QueryRowContext(ctx, q, append([]any{meetingID}, args...)...).Scan(&id)
	if errors.Is(err, sql.ErrNoRows) {
		return ErrNotFound
	}
	return err
}

const transcriptCols = "id, org_id, meeting_id, full_text, language, word_count, confidence_score, created_at, updated_at"

func scanTranscript(row interface{ Scan(...any) error }) (*model.Transcript, error) {
	var t model.Transcript
	err := row.Scan(&t.ID, &t.OrgID, &t.MeetingID, &t.FullText, &t.Language,
		&t.WordCount, &t.ConfidenceScore, &t.CreatedAt, &t.UpdatedAt)
	return &t, err
}

// GetTranscript returns the meeting's transcript (ports get_transcript). The
// meeting is first org-scoped (ErrNotFound on a missing/foreign meeting); a
// scoped meeting with no transcript row also yields ErrNotFound.
func GetTranscript(ctx context.Context, conn *sql.DB, p *Principal, meetingID string) (*model.Transcript, error) {
	if err := scopedMeetingID(ctx, conn, p, meetingID); err != nil {
		return nil, err
	}
	q := "SELECT " + transcriptCols + " FROM transcripts WHERE meeting_id = ?"
	t, err := scanTranscript(conn.QueryRowContext(ctx, q, meetingID))
	if errors.Is(err, sql.ErrNoRows) {
		return nil, ErrNotFound
	}
	return t, err
}

const segmentCols = "id, transcript_id, meeting_id, speaker_label, speaker_name, text, start_time, end_time, confidence, segment_index, is_partial, created_at, updated_at"

func scanSegment(row interface{ Scan(...any) error }) (*model.TranscriptSegment, error) {
	var s model.TranscriptSegment
	err := row.Scan(&s.ID, &s.TranscriptID, &s.MeetingID, &s.SpeakerLabel, &s.SpeakerName,
		&s.Text, &s.StartTime, &s.EndTime, &s.Confidence, &s.SegmentIndex, &s.IsPartial,
		&s.CreatedAt, &s.UpdatedAt)
	return &s, err
}

// SegmentFilters are the list query params (ports list_transcript_segments).
type SegmentFilters struct {
	Speaker string
	Q       string
	Limit   int
}

// ListTranscriptSegments returns the meeting's finalized (non-partial) segments
// ordered by start_time (ports list_transcript_segments). The meeting is first
// org-scoped (ErrNotFound on a missing/foreign meeting).
func ListTranscriptSegments(ctx context.Context, conn *sql.DB, p *Principal, meetingID string, f SegmentFilters) ([]model.TranscriptSegment, error) {
	if err := scopedMeetingID(ctx, conn, p, meetingID); err != nil {
		return nil, err
	}
	where := "meeting_id = ? AND is_partial = 0"
	args := []any{meetingID}
	if f.Speaker != "" {
		where += " AND speaker_label = ?"
		args = append(args, f.Speaker)
	}
	if f.Q != "" {
		where += " AND text LIKE ?"
		args = append(args, "%"+f.Q+"%")
	}
	limit := f.Limit
	if limit < 1 || limit > 500 {
		limit = 200
	}
	q := "SELECT " + segmentCols + " FROM transcript_segments WHERE " + where + " ORDER BY start_time LIMIT ?"
	args = append(args, limit)

	rows, err := conn.QueryContext(ctx, q, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]model.TranscriptSegment, 0)
	for rows.Next() {
		s, err := scanSegment(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, *s)
	}
	return out, rows.Err()
}

const participantCols = "id, meeting_id, speaker_label, speaker_name, user_id, email_address, joined_at, left_at, created_at, updated_at"

func scanParticipant(row interface{ Scan(...any) error }) (*model.MeetingParticipant, error) {
	var pt model.MeetingParticipant
	err := row.Scan(&pt.ID, &pt.MeetingID, &pt.SpeakerLabel, &pt.SpeakerName, &pt.UserID,
		&pt.EmailAddress, &pt.JoinedAt, &pt.LeftAt, &pt.CreatedAt, &pt.UpdatedAt)
	return &pt, err
}

// ListParticipants returns the meeting's participants, NULL joined_at last then
// by joined_at (ports list_participants). The meeting is first org-scoped.
func ListParticipants(ctx context.Context, conn *sql.DB, p *Principal, meetingID string) ([]model.MeetingParticipant, error) {
	if err := scopedMeetingID(ctx, conn, p, meetingID); err != nil {
		return nil, err
	}
	q := "SELECT " + participantCols + " FROM meeting_participants WHERE meeting_id = ? " +
		"ORDER BY (joined_at IS NULL), joined_at"
	rows, err := conn.QueryContext(ctx, q, meetingID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]model.MeetingParticipant, 0)
	for rows.Next() {
		pt, err := scanParticipant(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, *pt)
	}
	return out, rows.Err()
}

const chatCols = "id, meeting_id, org_id, sender_name, sender_email, text, sent_at, created_at"

func scanChatMessage(row interface{ Scan(...any) error }) (*model.MeetingChatMessage, error) {
	var c model.MeetingChatMessage
	err := row.Scan(&c.ID, &c.MeetingID, &c.OrgID, &c.SenderName, &c.SenderEmail,
		&c.Text, &c.SentAt, &c.CreatedAt)
	return &c, err
}

// ListChat returns the meeting's chat messages ordered by sent_at, capped at 500
// (ports list_chat). The meeting is first org-scoped.
func ListChat(ctx context.Context, conn *sql.DB, p *Principal, meetingID string) ([]model.MeetingChatMessage, error) {
	if err := scopedMeetingID(ctx, conn, p, meetingID); err != nil {
		return nil, err
	}
	q := "SELECT " + chatCols + " FROM meeting_chat_messages WHERE meeting_id = ? ORDER BY sent_at LIMIT 500"
	rows, err := conn.QueryContext(ctx, q, meetingID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]model.MeetingChatMessage, 0)
	for rows.Next() {
		c, err := scanChatMessage(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, *c)
	}
	return out, rows.Err()
}
